### PR TITLE
[SPARK-46213][PYTHON] Introduce `PySparkImportError` for error framework

### PR DESCRIPTION
**Pull request by itholic**
### What changes were proposed in this pull request?

This PR proposes to introduce `PySparkImportError` for error framework.

### Why are the changes needed?

For better error handling.

### Does this PR introduce _any_ user-facing change?

No API changes, but it's improve the user-facing error messages.

### How was this patch tested?

The existing CI should pass

### Was this patch authored or co-authored using generative AI tooling?

No.

### CODE REVIEW

**Review by dongjoon-hyun — APPROVED** [review]
In the PR description, `ImportError` instead of `IndexError`?
> replace IndexError in pyspark.sql.*.

For the code, +1, LGTM. Thank you, @itholic .

File: python/pyspark/errors/error_classes.py [hunk 1]
```diff
@@ -667,6 +667,11 @@
       "Only a single trigger is allowed."
     ]
   },
+  "PACKAGE_NOT_INSTALLED" : {
+    "message" : [
+      "<package_name> >= <minimum_version> must be installed; however, it was not found."
```
Comment (HyukjinKwon) [inline python/pyspark/errors/error_classes.py:672]: @itholic, the code here fails:

```python
>>> from pyspark.sql.pandas.utils import require_minimum_pyarrow_version
>>> require_minimum_pyarrow_version()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/.../spark/python/pyspark/sql/pandas/utils.py", line 68, in require_minimum_pyarrow_version
    raise PySparkImportError(
  File "/.../spark/python/pyspark/errors/exceptions/base.py", line 46, in __init__
    self._message = self._error_reader.get_error_message(
  File "/.../spark/python/pyspark/errors/utils.py", line 39, in get_error_message
    assert set(message_parameters_from_template) == set(message_parameters), (
AssertionError: Undefined error message parameter for error class: PACKAGE_NOT_INSTALLED. Parameters: {'package_name:': 'PyArrow', 'minimum_version': '4.0.0'}
```

This breaks scheduled Python build https://github.com/apache/spark/actions/runs/7088462269/job/19291103742. I an reverting this.

Comment (itholic) [reply python/pyspark/errors/error_classes.py:672]: Oh,,, seems like the special character `>` somehow breaks the build while parsing the error message format. Let me made a followup PR to fix it. Thanks for addressing this!

Comment (dongjoon-hyun) [reply python/pyspark/errors/error_classes.py:672]: Oh..

Comment (dongjoon-hyun) [reply python/pyspark/errors/error_classes.py:672]: Thank you for reverting.

**Pull request closed without merging**